smack-experimental/src/main/java/org/jivesoftware/smackx/mam/MamManager.java

```diff
@@ -0,0 +1,522 @@
+        RSMSet previousResultRsmSet = mamQueryResult.mamFin.getRSMSet();
+        RSMSet requestRsmSet = new RSMSet(count, previousResultRsmSet.getLast(), RSMSet.PageDirection.after);
+        return page(mamQueryResult.form, requestRsmSet);
+    }
+
+    private void preparePageQuery(MamQueryIQ mamQueryIQ, RSMSet rsmSet) {
+        mamQueryIQ.setType(IQ.Type.set);
+        mamQueryIQ.addExtension(rsmSet);
+    }
+
+    /**
+     * Get the form fields supported by the server.
+     * 
+     * @return the MAM query result.
+     * @throws NoResponseException
+     * @throws XMPPErrorException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     */
+    public MamQueryResult retrieveFormFields()
+            throws NoResponseException, XMPPErrorException, NotConnectedException, InterruptedException {
+        String queryId = UUID.randomUUID().toString();
+        MamQueryIQ mamQueryIQ = prepareMamQueryIQGet(queryId);
+        return queryArchive(mamQueryIQ, 0);
+    }
+
+    private MamQueryIQ prepareMamQueryIQSet(DataForm dataForm, String queryId) {
+        MamQueryIQ mamQueryIQ = new MamQueryIQ(queryId, dataForm);
+        mamQueryIQ.setType(IQ.Type.set);
+        return mamQueryIQ;
+    }
+
+    private MamQueryIQ prepareMamQueryIQGet(String queryId) {
+        MamQueryIQ mamQueryIQ = new MamQueryIQ(queryId, null);
+        mamQueryIQ.setType(IQ.Type.get);
+        return mamQueryIQ;
+    }
+
+    private MamQueryResult queryArchive(MamQueryIQ mamQueryIq, long extraTimeout)
+            throws NoResponseException, XMPPErrorException, NotConnectedException, InterruptedException {
+        if (extraTimeout < 0) {
+            throw new IllegalArgumentException("extra timeout must be zero or positive");
+        }
+
+        final XMPPConnection connection = connection();
+        MamFinExtension mamFinExtension = null;
+
+        PacketCollector finMessageCollector = connection.createPacketCollector(new MamMessageFinFilter(mamQueryIq));
+
+        PacketCollector.Configuration resultCollectorConfiguration = PacketCollector.newConfiguration()
+                .setStanzaFilter(new MamMessageResultFilter(mamQueryIq)).setCollectorToReset(finMessageCollector);
+
+        PacketCollector resultCollector = connection.createPacketCollector(resultCollectorConfiguration);
+
+        try {
+            connection.createPacketCollectorAndSend(mamQueryIq).nextResultOrThrow();
+            Message mamFinMessage = finMessageCollector
+                    .nextResultOrThrow(connection.getPacketReplyTimeout() + extraTimeout);
+            mamFinExtension = MamFinExtension.from(mamFinMessage);
+        } finally {
+            resultCollector.cancel();
+            finMessageCollector.cancel();
+        }
+
+        List<Forwarded> messages = new ArrayList<>(resultCollector.getCollectedCount());
+
+        for (Message resultMessage = resultCollector.pollResult(); resultMessage != null;) {
```

Comment:
This line leads to an endless loop, because only the first message is polled from the queue.
Quick fix:
for (Message resultMessage = resultCollector.pollResult(); resultMessage != null; resultMessage = resultCollector.pollResult()) {